src/flowgger/decoder/rfc3164_decoder.rs

```diff
@@ -92,22 +149,60 @@ fn parse_strip_pri(event: &str) -> Result<(Pri, &str), &'static str> {
     }
 }
 
-fn parse_ts(ts_str: &str) -> Result<f64, &'static str> {
-    // Append the year to parse a full ts
-    let current_year = Utc::now().year();
-    let ts = format!("{} {}", current_year, ts_str);
+fn parse_date_token<'a>(ts_tokens: &'a [&str]) -> Result<(f64, Vec<&'a str>), &'static str> {
+    // If we don't have at least 3 tokens, don't even try, parsing will fail
+    if ts_tokens.len() < 3 {
+        return Err("Invalid time format");
+    }
+    // Decode the date/time without year (expected), and if it fails, try  add the year
+    parse_date(ts_tokens, false).or_else(|_| parse_date(ts_tokens, true))
+}
 
-    match NaiveDateTime::parse_from_str(&ts, "%Y %b %d %H:%M:%S") {
-        Ok(date) => Ok(utils::PreciseTimestamp::from_naive_datetime(date).as_f64()),
-        Err(_) => Err("Unable to parse the date"),
+fn parse_date<'a>(ts_tokens: &'a [&str], has_year: bool) -> Result<(f64, Vec<&'a str>), &'static str> {
+    // Decode the date/time from the given tokens with optional year specified
+    let ts_str;
+    let mut idx;
+
+    // If no year in the string, parse manually add the current year
+    if has_year {
+        idx = 4;
+        ts_str = ts_tokens[0..idx].join(" ");
+    }
+    else {
+        idx = 3;
+        let current_year = Utc::now().year();
+        ts_str = format!("{} {}", current_year, ts_tokens[0..idx].join(" "));
+    }
+
+    match NaiveDateTime::parse_from_str(&ts_str, "%Y %b %d %H:%M:%S") {
+        Ok(naive_dt) => {
+            // See if the next token is a timezone
+            let mut ts = 0.0;
+            let tz_res: Result<Tz, String> = if ts_tokens.len() > idx { ts_tokens[idx].parse() } else { Err("No timezone".to_string()) };
+            if let Ok(tz) = tz_res {
+                let dt = tz.from_local_datetime(&naive_dt).single();
+                if dt.is_some() {
+                    ts = utils::PreciseTimestamp::from_datetime_tz(dt.unwrap()).as_f64();
+                    idx += 1;
+                }
+            }
+            // No timezome, give a timestamp without tz
+            else {
+                ts = utils::PreciseTimestamp::from_naive_datetime(naive_dt).as_f64();
+            }
+            Ok((ts, ts_tokens[idx..].to_vec()))
+        },
+        Err(_err) => {
+            Err("Unable to parse date")
```

Comment:
Minor: I would still put the error message from parse_from_string at the end of the custom message:

```rust
Err(format!("Unable to parse date: {}", err))
```

Reply:
I'd love to, i tried on my very first flowgger updates but the whole app is build on using Err with static strings. I would require an update of almost every module. I have future plans for doing that when improving the log system and adding metrics hook.

Reply:
Ahhh I remember now..